lib/imgresizer.js

```diff
@@ -147,64 +154,49 @@ function processImage(params, method, callback) {
-
-                  log.verbose('the image "' + path.basename(dest) + '" seems to be already converted and cached, skipping...');
-                  callback(null, dest, info);
-                });
-              });
-            });
+          // We already have a cached transform; just return image info.
+          imagemagick.info(dest, function(err, info) {
+            if (err) {
+              logError(err);
+              callback(err);
+              return;
+            }
+            log.verbose('The image "' + path.basename(dest) + '" is already converted and cached, skipping...');
+            callback(null, dest, info);
           });
         } else {
-          // prepare file for converting, by opening it in write (append) mode
-          fs.open(dest, 'a', function(err, fd) {
-            // try to aquire EXCLUSIVE lock (cause we're writing to file)
-            fs.flock(fd, 'exnb', function(err) {
+
+          // Transform the image; start in working dir and promote to cache dir once transform succeeds.
+          imgcache.get_working_path(filename, function(workingPath, workingExists) {
+            imgcache.ensure_path(workingPath, function(err) {
               if (err) {
-                log.trace('couldn\'t lock ' + dest + ', while converting (' + err.message + ')');
-                return fs.close(fd, function() {
-                  process.nextTick(function() {
-                    processImage(params, method, callback);
-                  }); // retry
-                });
+                logError(err);
+                callback(err);
+                return;
               }
 
-              // just log it, useful when debugging
-              log.verbose('locked ' + dest + ', starting converting...');
+              log.verbose('Transforming... src=' + downloaded_image + ', dst=' + workingPath);
 
               imagemagick[method]({
                 src: downloaded_image,
-                dst: dest,
+                dst: workingPath,
                 width: width,
                 height: height,
                 quality: quality
               }, force, function(err, info) {
-                fs.close(fd, function() {
+                if (err) {
+                  logError(err);
+                  callback(err);
+                  return;
+                }
+                imgcache.promote_working_path_to_cache(workingPath, function(err, cachePath) {
                   if (err) {
-                    fs.unlink(dest);
-                    log.error(err.stack);
-                    return callback(err);
+                    logError(err);
+                    callback(err);
+                    return;
                   }
-
-                  log.info('converted image: ' + downloaded_image + ', and saved to: ' + dest);
-                  callback(null, dest, info);
+                  // Success! Return the full path to cached transform result and image info.
+                  callback(null, cachePath, info);
```

Comment:
NIT: seems like there is a bit of callback nesting here. Promises or an async.waterfall might make it a little easier to look at. But functionally looks good.